src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/TestHooks/ValueTaskExtensions.cs

```diff
@@ -0,0 +1,255 @@
+                _task = task;
+                _captureContext = captureContext;
+            }
+
+            /// <summary>
+            /// Gets the awaiter.
+            /// </summary>
+            /// <returns>The awaiter.</returns>
+            public NoThrowValueTaskAwaiter<TResult> GetAwaiter()
+            {
+                return new NoThrowValueTaskAwaiter<TResult>(_task, _captureContext);
+            }
+        }
+
+        /// <summary>
+        /// An awaiter that wraps a task and never throws an exception when waited on.
+        /// </summary>
+        public readonly struct NoThrowValueTaskAwaiter<TResult> : ICriticalNotifyCompletion
+        {
+            /// <summary>
+            /// The task.
+            /// </summary>
+            private readonly ValueTask<TResult> _task;
+
+            /// <summary>
+            /// A value indicating whether the continuation should be scheduled on the current sync context.
+            /// </summary>
+            private readonly bool _captureContext;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="NoThrowValueTaskAwaiter{TResult}"/> struct.
+            /// </summary>
+            /// <param name="task">The task.</param>
+            /// <param name="captureContext">if set to <c>true</c> [capture context].</param>
+            public NoThrowValueTaskAwaiter(ValueTask<TResult> task, bool captureContext)
+            {
+                Contract.ThrowIfNull(task, nameof(task));
+                _task = task;
+                _captureContext = captureContext;
+            }
+
+            /// <summary>
+            /// Gets a value indicating whether the task has completed.
+            /// </summary>
+            public bool IsCompleted
+            {
+                get { return _task.IsCompleted; }
+            }
+
+            /// <summary>
+            /// Schedules a delegate for execution at the conclusion of a task's execution.
+            /// </summary>
+            /// <param name="continuation">The action.</param>
+            public void OnCompleted(Action continuation)
+            {
+                _task.ConfigureAwait(_captureContext).GetAwaiter().OnCompleted(continuation);
+            }
+
+            /// <summary>
+            /// Schedules a delegate for execution at the conclusion of a task's execution
+            /// without capturing the ExecutionContext.
+            /// </summary>
+            /// <param name="continuation">The action.</param>
+            public void UnsafeOnCompleted(Action continuation)
+            {
+                _task.ConfigureAwait(_captureContext).GetAwaiter().UnsafeOnCompleted(continuation);
+            }
+
+            /// <summary>
+            /// Does nothing.
+            /// </summary>
+            public (TaskStatus Status, ExceptionDispatchInfo? Exception, TResult? Result) GetResult()
+            {
+                var preserved = _task.Preserve();
```

Comment:
If this VT is backed by an IValueTaskSource and it's faulted or canceled, this is going to result in throwing and catching its exception. My understanding is that's the cost this whole thing is trying to prevent.

Reply:
This is also on very shaky ground. It may work most of the time, but again with an IVTS, this is calling Preserve after it's already hooked up a continuation to the IVTS. That's not something that's technically supported, and could possibly fail with some implementations. I _think_ it should work with ours. But, shaky ground.